README
NAME

    Log::Any::Plugin - Adapter-modifying plugins for Log::Any

VERSION

    version 0.003

SYNOPSIS

        use Log::Any::Adapter;
        use Log::Any::Plugin;
    
        # Create your adapter as normal
        Log::Any::Adapter->set( 'SomeAdapter' );
    
        # Add plugin to modify its behaviour
        Log::Any::Plugin->add( 'Stringify' );
    
        # Multiple plugins may be used together
        Log::Any::Plugin->add( 'Levels', level => 'debug' );

DESCRIPTION

    Log::Any::Plugin is a method for augmenting arbitrary instances of
    Log::Any::Adapters.

    Log::Any::Plugins work much in the same manner as Moose 'around'
    modifiers to augment logging behaviour of pre-existing adapters.

MOTIVATION

    Many of the Log::Any::Adapters have extended functionality, such as
    being able to selectively disable various log levels, or to handle
    multiple arguments.

    In order for Log::Any to be truly 'any', only the common subset of
    adapter functionality can be used. Any specific adapter functionality
    must be avoided if there is a possibility of using a different adapter
    at a later date.

    Log::Any::Plugins provide a method to augment adapters with missing
    functionality so that a superset of adapter functionality can be used.

METHODS

 add ( $plugin, [ %plugin_args ] )

    This is the single method for adding plugins to adapters. It works in a
    similar function to Log::Any::Adapter->set()

      * $plugin

      The plugin class to add to the currently active adapter. If the class
      is in the Log::Any::Plugin:: namespace, you can simply specify the
      name, otherwise prefix a '+'.

          eg. '+My::Plugin::Class'

      * %plugin_args

      These are plugin specific arguments. See the individual plugin
      documentation for what options are supported.

SEE ALSO

    Log::Any, Log::Any::Plugin::Levels, Log::Any::Plugin::Stringify

ACKNOWLEDGEMENTS

    Thanks to Strategic Data for sponsoring the development of this module.

AUTHOR

    Stephen Thirlwall <sdt@cpan.org>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2014 by Stephen Thirlwall.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision History for perl module Log::Any::Plugin

0.003     2014-10-09T10:31:03AEDT+1100

    - Remove missing Log::Any::Adapter::Util dependency
        - fixes RT:99373 - thanks Mark Lawrence

0.002     2013-08-30T16:44:53EST+1000

    - Performance improvements
    - Remove before/after/around functions
        - use Class::Method::Modifiers instead if you need these

0.001     2011-09-06

    - Initial release


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "English" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "Log::Any::Adapter" => 0,
    "Module::Build" => "0.3601",
    "Test::Differences" => 0,
    "Test::Exception" => 0,
    "Test::More" => 0,
    "Test::NoWarnings" => 0
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Adapter-modifying plugins for Log::Any",
  "dist_author" => [
    "Stephen Thirlwall <sdt\@cpan.org>"
  ],
  "dist_name" => "Log-Any-Plugin",
  "dist_version" => "0.001",
  "license" => "perl",
  "module_name" => "Log::Any::Plugin",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "Exporter" => 0,
    "Hash::Util" => 0,
    "Log::Any" => "0.12",
    "Log::Any::Adapter::Util" => 0,
    "base" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Log/Any/Plugin.pm
package Log::Any::Plugin;
# ABSTRACT: Adapter-modifying plugins for Log::Any
$Log::Any::Plugin::VERSION = '0.003';
use strict;
use warnings;

use Log::Any 0.12;
use Log::Any::Plugin::Util  qw( get_class_name  );

use Carp qw( croak );

sub add {
    my ($class, $plugin_class, %plugin_args) = @_;

    my $adapter_class = ref Log::Any->get_logger(category => caller());

    $plugin_class = get_class_name($plugin_class);
    unless ( defined( eval "require $plugin_class" ) )
    {    ## no critic (ProhibitStringyEval)
        die $@;
    }

    $plugin_class->install($adapter_class, %plugin_args);
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Log::Any::Plugin - Adapter-modifying plugins for Log::Any

=head1 VERSION

version 0.003

=head1 SYNOPSIS

    use Log::Any::Adapter;
    use Log::Any::Plugin;

    # Create your adapter as normal
    Log::Any::Adapter->set( 'SomeAdapter' );

    # Add plugin to modify its behaviour
    Log::Any::Plugin->add( 'Stringify' );

    # Multiple plugins may be used together
    Log::Any::Plugin->add( 'Levels', level => 'debug' );

=head1 DESCRIPTION

Log::Any::Plugin is a method for augmenting arbitrary instances of
Log::Any::Adapters.

Log::Any::Plugins work much in the same manner as Moose 'around' modifiers to
augment logging behaviour of pre-existing adapters.

=head1 MOTIVATION

Many of the Log::Any::Adapters have extended functionality, such as being
able to selectively disable various log levels, or to handle multiple arguments.

In order for Log::Any to be truly 'any', only the common subset of adapter
functionality can be used. Any specific adapter functionality must be avoided
if there is a possibility of using a different adapter at a later date.

Log::Any::Plugins provide a method to augment adapters with missing
functionality so that a superset of adapter functionality can be used.

=head1 METHODS

=head2 add ( $plugin, [ %plugin_args ] )

This is the single method for adding plugins to adapters. It works in a
similar function to Log::Any::Adapter->set()

=over

=item * $plugin

The plugin class to add to the currently active adapter. If the class is in
the Log::Any::Plugin:: namespace, you can simply specify the name, otherwise
prefix a '+'.

    eg. '+My::Plugin::Class'

=item * %plugin_args

These are plugin specific arguments. See the individual plugin documentation for
what options are supported.

=back

=head1 SEE ALSO

L<Log::Any>, L<Log::Any::Plugin::Levels>, L<Log::Any::Plugin::Stringify>

=head1 ACKNOWLEDGEMENTS

Thanks to Strategic Data for sponsoring the development of this module.

=head1 AUTHOR

Stephen Thirlwall <sdt@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Stephen Thirlwall.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Log/Any/Plugin/Util.pm
package Log::Any::Plugin::Util;
# ABSTRACT: Utilities for Log::Any::Plugin classes
$Log::Any::Plugin::Util::VERSION = '0.003';
use strict;
use warnings;
use Carp qw(croak);

use base qw(Exporter);

our @EXPORT_OK = qw(
    get_old_method
    set_new_method
    get_class_name
);

sub get_old_method {
    my ($class, $method_name) = @_;
    return $class->can($method_name);
}

sub set_new_method {
    my ($class, $method_name, $new_method) = @_;

    no warnings 'redefine';
    no strict 'refs'; ## no critic (ProhibitNoStrict)
    *{ $class . '::' . $method_name } = $new_method;
}

sub get_class_name {
    my ($name) = @_;

    return substr($name, 0, 1) eq '+' ? substr($name, 1)
                                      : 'Log::Any::Plugin::' . $name;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Log::Any::Plugin::Util - Utilities for Log::Any::Plugin classes

=head1 VERSION

version 0.003

=head1 DESCRIPTION

These functions are only of use to authors of Log::Any::Plugin classes.

Users should see Log::Any::Plugin instead.

=head1 FUNCTIONS

=head2 get_old_method ( $class, $method_name )

Returns a coderef of the existing method in the class, or undef if none exists.
Exactly the same semantics as $class->can($method_name).

=head2 set_new_method ( $class, $method_name, &new_method )

Replaces the given method with the new version.

=over

=item * $class

Name of class containing the method.

=item * $method_name

Name of method to be modified.

=item * &new_method

Coderef of the new method.

=back

=head2 get_class_name ( $name )

Creates a fully-qualified class name from the abbreviated class name rules
in Log::Any::Plugin.

=over

=item * $name

Either a namespace suffix, or a fully-qualified class name prefixed with '+'.

=back

=head1 SEE ALSO

L<Log::Any::Plugin>

=head1 ACKNOWLEDGEMENTS

Thanks to Strategic Data for sponsoring the development of this module.

=head1 AUTHOR

Stephen Thirlwall <sdt@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Stephen Thirlwall.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Log/Any/Plugin/Levels.pm
package Log::Any::Plugin::Levels;
# ABSTRACT: Logging-level filtering plugin for log adapters
$Log::Any::Plugin::Levels::VERSION = '0.003';
use strict;
use warnings;
use Carp qw(croak);
use Hash::Util qw( lock_hash );
use Log::Any;

use Log::Any::Plugin::Util qw( get_old_method set_new_method );

my $level_count = 1;
my %level_value = map { $_ => $level_count++ } Log::Any->logging_methods();
my %level_name = reverse %level_value;
$level_value{all} = 1;
lock_hash(%level_value);

my $default_value = $level_value{warning};

# Inside-out storage for level field.
my %selected_level_value;
my %selected_level_name;

sub install {
    my ($class, $adapter_class, %args) = @_;

    my $accessor = $args{accessor} || 'level';
    croak $adapter_class . '::' . $accessor
        . q( already exists - use 'accessor' to specify another method name)
        if get_old_method($adapter_class, $accessor);

    if ($args{level}) {
        $default_value = _get_level_value($args{level});
    }

    # Create the $log->level accessor
    set_new_method($adapter_class, $accessor, sub {
        my $self = shift;
        if (@_) {
            my $level_name = shift;
            $selected_level_value{$self} = _get_level_value($level_name);
            $selected_level_name{$self} = $level_name;
        }
        return $selected_level_name{$self};
    });

    # Augment the $log->debug methods
    for my $method_name ( Log::Any->logging_methods() ) {
        my $level = $level_value{$method_name};

        my $old_method = get_old_method($adapter_class, $method_name);
        set_new_method($adapter_class, $method_name, sub {
            my $self = shift;
            return if ($selected_level_value{$self} || $default_value) > $level;
            $self->$old_method(@_);
        });
    }

    # Augment the $log->is_debug methods
    for my $method_name ( Log::Any->detection_methods() ) {
        my $level = $method_name;
        $level =~ s/^is_//;
        $level = $level_value{$level};

        my $old_method = get_old_method($adapter_class, $method_name);
        set_new_method($adapter_class, $method_name, sub {
            my $self = shift;
            return (($selected_level_value{$self} || $default_value) <= $level)
                && $self->$old_method(@_);
        });
    }
}

sub _get_level_value {
    my ($level_name) = @_;
    return $default_value if ($level_name eq 'default');
    croak('Unknown log level ' . $level_name)
        unless exists $level_value{$level_name};
    return $level_value{$level_name};
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Log::Any::Plugin::Levels - Logging-level filtering plugin for log adapters

=head1 VERSION

version 0.003

=head1 SYNOPSIS

    # Set up some kind of logger.
    use Log::Any::Adapter;
    Log::Any::Adapter->set('SomeAdapter');

    # Apply the Levels plugin to your logger
    use Log::Any::Plugin;
    Log::Any::Plugin->add('Levels', level => 'debug');


    # In your modules
    use Log::Any qw($log);

    $log->trace('trace'); # this log is ignored
    $log->error('error'); # this log gets through

    $log->level('trace');
    $log->trace('trace'); # this gets through now

=head1 DESCRIPTION

Log::Any leaves the decision of which log levels to ignore and which to
actually log down to the individual adapters. Many adapters simply log
everything.

Log::Any::Plugin::Levels allows you to add level filtering functionality into
any adapter. Logs lower than $log->level are ignored.

The $log->is_debug family of functions are modified to reflect this level.

=head1 NAME

Log::Any::Plugin::FilterArgs - custom log-level filtering for log adapters

=head1 CONFIGURATION

Configuration values are passed as key-value pairs when adding the plugin:
    Log::Any::Plugin->add('Levels',
                            level    => 'debug',
                            accessor => 'my_level');

=head2 level => $default_level

The global log level, which defaults to 'warning'. See the level method below
for a discussion on how this is applied.

=head2 accessor => $accessor_name

This is the name of the $log->level accessor function.

The default value is 'level'. This can be changed to avoid any name clashes
that may occur. An exception will be thrown in the case of a name clash.

=head1 METHODS

There are no methods in this package which should be directly called by the
user. Use Log::Any::Plugin->add() instead.

=head2 install

Private method called by Log::Any::Plugin->add()

=head1 ADAPTER METHODS

The following methods are injected into the adapter class.

=head2 level( [ $log_level ] )

Accessor for the current log level in the calling $log object.

All $log objects start with the default level specified when adding the
plugin.  Individual $log objects can set a custom level with this accessor.

To reset to the default log level, specify 'default'.

=head1 SEE ALSO

L<Log::Any::Plugin>

=head1 ACKNOWLEDGEMENTS

Thanks to Strategic Data for sponsoring the development of this module.

=head1 AUTHOR

Stephen Thirlwall <sdt@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Stephen Thirlwall.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Log/Any/Plugin/Stringify.pm
package Log::Any::Plugin::Stringify;
# ABSTRACT: Custom argument stringification plugin for log adapters
$Log::Any::Plugin::Stringify::VERSION = '0.003';
use strict;
use warnings;

use Log::Any::Plugin::Util qw( get_old_method set_new_method );

use Data::Dumper;

sub install {
    my ($class, $adapter_class, %args) = @_;

    my $stringifier = $args{stringifier} || \&default_stringifier;

    # Inject the stringifier into the existing logging methods
    #
    for my $method_name ( Log::Any->logging_methods() ) {
        my $old_method = get_old_method($adapter_class, $method_name);
        set_new_method($adapter_class, $method_name, sub {
            my $self = shift;
            $self->$old_method($stringifier->(@_));
        });
    }
}

sub default_stringifier {
    my (@args) = @_;

    local $Data::Dumper::Indent    = 0;
    local $Data::Dumper::Pair      = '=';
    local $Data::Dumper::Quotekeys = 0;
    local $Data::Dumper::Sortkeys  = 1;
    local $Data::Dumper::Terse     = 1;

    return join('', map { ref $_ ? Dumper($_) : $_ } @args);
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Log::Any::Plugin::Stringify - Custom argument stringification plugin for log adapters

=head1 VERSION

version 0.003

=head1 SYNOPSIS

    # Set up some kind of logger
    use Log::Any::Adapter;
    Log::Any::Adapter->set('SomeAdapter');

    # Apply your own argument stringifier.
    use Log::Any::Plugin;
    Log::Any::Plugin->add('Stringify', \&my_stringifier);

=head1 DESCRIPTION

Log::Any logging functions are only defined to have a single $msg argument.
Some adapters accept multiple arguments (like print does), but many don't.
You may also want to do some sort of stringification of hash and list refs.

Log::Any::Plugin::Stringify allows you to inject an argument stringification
function into every logging call, so that when you write this:

    $log->error( ... );

you effectively get this:

    $log->error( my_function( ... ) );

=head1 CONFIGURATION

These configuration values are passed as key-value pairs:
    Log::Any::Plugin->add('Stringify', stringifier => \&my_func);

=head2 stringifier => &my_func

The stringifier function takes a list of arguments and should return a single
string.

See default_stringifier below for the default stringifier.

=head1 METHODS

There are no methods in this package which should be directly called by the
user.  Use Log::Any::Plugin->add() instead.

=head2 install

Private method called by Log::Any::Plugin->add()

=head2 default_stringifier

The default stringifier function if none is supplied. Listrefs and hashrefs are
expanded by Data::Dumper, and the whole lot is concatenated into one string.

=head1 SEE ALSO

L<Log::Any::Plugin>

=head1 ACKNOWLEDGEMENTS

Thanks to Strategic Data for sponsoring the development of this module.

=head1 AUTHOR

Stephen Thirlwall <sdt@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Stephen Thirlwall.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Adapter-modifying plugins for Log::Any",
  "AUTHOR" => "Stephen Thirlwall <sdt\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Log-Any-Plugin",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Log::Any::Plugin",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "Exporter" => 0,
    "Hash::Util" => 0,
    "Log::Any" => "0.12",
    "base" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Log::Any::Adapter" => 0,
    "Test::Differences" => 0,
    "Test::Exception" => 0,
    "Test::More" => 0,
    "Test::NoWarnings" => 0
  },
  "VERSION" => "0.003",
  "test" => {
    "TESTS" => "t/*.t t/plugin/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Data::Dumper" => 0,
  "Exporter" => 0,
  "File::Spec" => 0,
  "Hash::Util" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Log::Any" => "0.12",
  "Log::Any::Adapter" => 0,
  "Test::Differences" => 0,
  "Test::Exception" => 0,
  "Test::More" => 0,
  "Test::NoWarnings" => 0,
  "base" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



